### PR TITLE
posix: implement sys_futimes() (previously sys_utimes())

### PR DESCRIPTION
**Pull request by ziemleszcz**
JIRA: DTR-276

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: (https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/94).
- [x] Tested by hand on: (imx6ull).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Comment by ziemleszcz** [comment]
sys_utimes to be removed later (syscall count will not change).

File: posix/posix.c [hunk 1]
```diff
@@ -1993,14 +1993,45 @@ int posix_setsockopt(int socket, int level, int optname, const void *optval, soc
...
+		return err;
+
+	hal_memset(&msg, 0, sizeof(msg_t));
+
+	msg.type = mtSetAttr;
+	hal_memcpy(&msg.i.attr.oid, &f->oid, sizeof(oid_t));
+
+	msg.i.attr.type = atMTime;
+	msg.i.attr.val = times[1].tv_sec;
+	if (((err = proc_send(f->oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
```
Comment (nalajcie) [inline posix/posix.c:2021]: missing `flieDeref`

File: posix/posix.c [hunk 2]
```diff
@@ -1993,14 +1993,45 @@ int posix_setsockopt(int socket, int level, int optname, const void *optval, soc
...
+		return err;
+
+	msg.i.attr.type = atATime;
+	msg.i.attr.val = times[0].tv_sec;
+	if (((err = proc_send(f->oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+		return err;
```
Comment (nalajcie) [inline posix/posix.c:2027]: missing `flieDeref`

File: include/syscalls.h [hunk 3]
```diff
@@ -108,6 +108,7 @@
 	\
 	ID(sys_ioctl) \
 	ID(sys_utimes) \
+	ID(sys_futimes) \
```
Comment (nalajcie) [inline include/syscalls.h:111]: maybe we can remove `sys_utimes` now that it's implemented in userspace?

File: posix/posix.c [hunk 4]
```diff
@@ -1993,14 +1993,45 @@ int posix_setsockopt(int socket, int level, int optname, const void *optval, soc
...
+int posix_futimes(int fildes, const struct timeval *times)
+{
+	TRACE("futimes(%d)", fildes);
+
+	open_file_t *f;
+	msg_t msg;
+	int err;
+
+	if ((err = posix_getOpenFile(fildes, &f)) < 0)
+		return err;
```
Comment (nalajcie) [inline posix/posix.c:2012]: maybe we should guard against `times` being null (somebody calling syscall directly) and return error (`futimes(_, NULL)` function call handled correctly in libphoenix)

**Review by nalajcie — CHANGES_REQUESTED** [review]
IMHO the decision regarding removing `utimes` syscall should be taken by @agkaminski.

nitpicking: `futimes` is non-posix (but exists on BSD also)... alas `futimens` (nanosecond precision variant) is specified in POSIX.1-2008. Maybe we should use this syscall as a part of 
POSIX compat layer (even if it would support only seconds - precision for now)?

File: include/syscalls.h (hunk 3 above)
Comment (ziemleszcz) [reply include/syscalls.h:111]: I will do it in separate pull request to not break libphoenix building.

File: posix/posix.c (hunk 4 above)
Comment (ziemleszcz) [reply posix/posix.c:2012]: We can assume that this pointer should not be NULL.

File: include/syscalls.h (hunk 3 above)
Comment (nalajcie) [reply include/syscalls.h:111]: ack

File: posix/posix.c (hunk 4 above)
Comment (agkaminski) [reply posix/posix.c:2012]: @nalajcie We should really sanitize all pointers passed to the kernel - not only check if it's not NULL but also if it belongs to the caller's map

Comment (ziemleszcz) [reply posix/posix.c:2012]: Agreed but this applies to all syscalls so we should have some common sanitizing code.

**Comment by agkaminski** [comment]
@nalajcie change from utimes to futimes is ok - kernel knows about file descriptors, everything can be done in user space. Do we have already clear goal of specific POSIX version compliance? If we want to introduce nanosecond precision much of existing API need to change - imho thats something to do in the future, for now futimes is ok to me.

**Comment by nalajcie** [comment]
> @nalajcie change from utimes to futimes is ok - kernel knows about file descriptors, everything can be done in user space. Do we have already clear goal of specific POSIX version compliance? If we want to introduce nanosecond precision much of existing API need to change - imho thats something to do in the future, for now futimes is ok to me.

It's OK for now pro me also, just doesn't sit well with me that we're adding non-posix function to posix.c :).

FYI, on phoenix-rtos.com webpage it's written that we're aiming for  IEEE Std 1003.1-2017 (so POSIX.1-2017) compliance (but we may change that ofc).

Other than that, LGTM (but assuming @agkaminski should give approval :man_mechanic:)

**Review by agkaminski — APPROVED** [review]

File: posix/posix.c [hunk 5]
```diff
@@ -1993,14 +1993,60 @@ int posix_setsockopt(int socket, int level, int optname, const void *optval, soc
...
+	msg.i.attr.val = times[1].tv_sec;
+	err = proc_send(f->oid.port, &msg);
+	if (err < 0) {
+		posix_fileDeref(f);
+		return err;
+	}
+	if (msg.o.attr.err < 0) {
+		posix_fileDeref(f);
+		return msg.o.attr.err;
+	}
```
Comment (kemonats) [inline posix/posix.c:2030]: You can simplify it a litle bit.
```suggestion
	if ((err >= 0) && (msg.o.attr.err < 0)) {
		err = msg.o.attr.err;
	}
	if (err < 0) {
		posix_fileDeref(f);
		return err;
	}
```

File: posix/posix.c [hunk 6]
```diff
@@ -1993,14 +1993,60 @@ int posix_setsockopt(int socket, int level, int optname, const void *optval, soc
...
+	msg.i.attr.val = times[0].tv_sec;
+	err = proc_send(f->oid.port, &msg);
+	if (err < 0) {
+		posix_fileDeref(f);
+		return err;
+	}
+	if (msg.o.attr.err < 0) {
+		posix_fileDeref(f);
+		return msg.o.attr.err;
+	}
```
Comment (kemonats) [inline posix/posix.c:2042]: and this
```suggestion
	if (err >= 0) {
		err = EOK;
		if (msg.o.attr.err < 0) {
			err = msg.o.attr.err;
		}
	}
```

**Review by nalajcie — APPROVED** [review]

**Pull request merged**